### PR TITLE
LPS-117986 Fixes query selection in CKEditor autocomplete plugin

### PR DESCRIPTION
**Pull request by jbalsas**
Revamp of https://github.com/liferay-frontend/liferay-portal/pull/204

> I haven't had a lot of time to look into this, but from what I've seen, the issue is related to [#10031 CKEditor injects special character in Chrome](https://dev.ckeditor.com/ticket/10031). This is to be expected on Webkit-based browsers and CKEditor accounts for in its [FILLING_CHAR_SEQUENCE](https://ckeditor.com/docs/ckeditor4/latest/api/CKEDITOR_dom_selection.html#property-FILLING_CHAR_SEQUENCE)
>
> What I'm seeing is that in presence of the filler char, our computation is off by as many filler chars as present, making the query fail. Curiously enough, an arrow movement (back) or mouse click make the fillers go away and allow the autocomplete to show again.

As stated in the commit message:

Depending on the scenario, the character before the last trigger can be:
- A space
- An `&nbsp;` entity
- A filler sequence (7 `\u200b` chars in webkit browsers)

This PR simply makes sure that all 3 are detected as limits for the query sequence.

This should fix:
- [LPP-38325 Cannot add 2 consecutive mentions](https://issues.liferay.com/browse/LPP-38325) (positive)
- [LPP-38323 Second mention overwrites the first one](https://issues.liferay.com/browse/LPP-38323) (hopeful)

@antonio-ortega, @holatuwol, @hongvo2308, could you test this solution to see if it works as you'd expect? As I said, I didn't carry over the previous modification since I wasn't sure what was its intended effect. Please, feel free to clarify and work with someone on the team to add it to the PR if indeed necessary!

### CODE REVIEW

**Comment by wincent** [comment]
ci:test:sf

**Comment by wincent** [comment]
ci:test:relevant

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js [hunk 1]
```diff
@@ -183,6 +183,25 @@
 				}
 			});
 
+			// What follows is an algorithm to properly select the query from the
+			// last detected trigger. Because of the HTML structure, it's not as
+			// straightforward as a text search.
+			//
+			// If triggerIndex === -1, the trigger is not in the current html node
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js:190]: html → HTML

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js [hunk 2]
```diff
@@ -183,6 +183,25 @@
...
+			// element. Thus, we walk the DOM tree upwards until we find it,
+			// constructing the query as we visit the nodes in the tree.
+			//
+			// If triggerIndex > 1, we found the trigger in a longer text sequence.
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js:194]: `triggerIndex > 1` → `triggerIndex > 0`

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js [hunk 3]
```diff
@@ -183,6 +183,25 @@
...
+			// We check if the char to the trigger is a space (' ' or nbsp;) by
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js:195]: "if the char to the trigger" → "if the char before the trigger"

`nbsp;` → `&nbsp;`

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js [hunk 4]
```diff
@@ -183,6 +183,25 @@
...
+			// checking assserting that trimming the char returns false or a
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js:196]: "checking asserting" → "checking"

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js [hunk 5]
```diff
@@ -183,6 +183,25 @@
...
+			// filler char (\u200b) for webkit-based engines. If that's the case,
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js:197]: webkit → WebKit

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js [hunk 6]
```diff
@@ -183,6 +183,25 @@
...
+			// we take the substring from triggerPosition forward and discard
+			// the rest of the text sequence.
+			//
+			// If triggerIndex === 1, the trigger is the first char in the sequence
```
Comment (wincent) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js:201]: `triggerIndex === 1` → `triggerIndex === 0`

**Comment by wincent** [comment]
Test failure looks unrelated:

```
clean-up-java-processes:
   [delete] Deleting: /opt/dev/projects/github/liferay-portal/clean_up_java_processes.sh
     ...
     [echo] [Baseline Warning] Bundle Version Change Recommended: 5.13.3
```

Going to re-run anyway to see if we can get a green run.

**Comment by wincent** [comment]
ci:test:relevant

**Comment by hongvo2308** [comment]
Hi, I have tested this implementation.
It could not solve the problem in the https://issues.liferay.com/browse/LPP-38325.
I think we need to consider making https://github.com/liferay-frontend/liferay-portal/blob/463374273efaed0b1324de047542fce1587ccfe5/modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/autocomplete/autocomplete.js#L386 execute.
Thank you.

**Comment by wincent** [comment]
Ok, thanks for testing @hongvo2308. I've force-pushed to include [your original fix](https://github.com/liferay-frontend/liferay-portal/pull/223/commits/635279aa92b0174828ff319857b6c4ebb3248a5e) as well as [Chema's change](https://github.com/liferay-frontend/liferay-portal/pull/223/commits/4e23774209470313518596d13f8b5e40120ba512). I also [made the copy edits](https://github.com/liferay-frontend/liferay-portal/pull/223/commits/17e1818cdc8237e42a6ddfbc88f42b58f1c5850c) that I had commented on above.

Now, I haven't repeated the analysis that you all have done, so I have no idea why we have the contradiction here...

> I don't see how this change can actually fix the reported issue... the change is done in the method that's used to replace the html once an item from the autocomplete list is selected. The reported issue states that the list doesn't show up the second time, so that conditional should have no effect whatsoever on this. ([comment](https://github.com/liferay-frontend/liferay-portal/pull/204#issuecomment-670694644))

vs

> I have tested this implementation.
> It could not solve the problem ([comment](https://github.com/liferay-frontend/liferay-portal/pull/223#issuecomment-671714175))

🤷‍♂️ 

So let's try one more time and see whether both of these get fixed by the combined change:

> This should fix:
> 
> * [LPP-38325 Cannot add 2 consecutive mentions](https://issues.liferay.com/browse/LPP-38325) (positive)
> * [LPP-38323 Second mention overwrites the first one](https://issues.liferay.com/browse/LPP-38323) (hopeful)

I'll kick off the automated tests, but can you please verify that this works in its combined form?

**Comment by wincent** [comment]
ci:test:sf

**Comment by wincent** [comment]
ci:test:relevant

**Comment by wincent** [comment]
ci:test:bundle

**Comment by jbalsas** [comment]
Thanks @wincent!

> Now, I haven't repeated the analysis that you all have done, so I have no idea why we have the contradiction here...

- I took this from https://github.com/liferay-frontend/liferay-portal/pull/204 
- Which points to [Cannot add 2 consecutive mentions](https://issues.liferay.com/browse/LPP-38325)

Based on the acceptance Criteria:
> **Actual result**
> Mention list is not displayed.
>
> **Expected result**
> Mention list is displayed and can be added successfully.

So, what I failed to see is how the fix would address the first item "mention list is displayed". I honestly failed to register the "can be added successfully" part of the equation myself.

That being said, I did check and verified this worked (at least on my machine) on Chrome, which is why I said I was "hopeful" adding multiple items should now work successfully. Maybe the fix is needed in a different situation? or browser?

I tried to make sense of the comments in https://github.com/holatuwol/liferay-portal/pull/203#issuecomment-665452319 but it felt like there was some missing information coming from some other conversations that I couldn't track down. Maybe @hongvo2308, @holatuwol or @antonio-ortega can share their analysis so we can continue from there.

**Comment by hongvo2308** [comment]
Hi @wincent  @jbalsas ,
I am sorry, I have mistaken betwwen https://issues.liferay.com/browse/LPP-38323 and  https://issues.liferay.com/browse/LPP-38325. I have tested on the  https://issues.liferay.com/browse/LPP-38323 for the last time but I wrote  https://issues.liferay.com/browse/LPP-38325.
About my analysis for this one is that the split function should executed so that the replacement will take place at the right position which is found by the previousTrigger.
Sorry for this mistake.

**Comment by hongvo2308** [comment]
Hi @wincent @jbalsas 
It has been 6 days since the last update. So I would like to ask if there is any update on this ticket.
I have tested again and I could confirm that the @jbalsas 's implementation solves the issue https://issues.liferay.com/browse/LPP-38325. And the current implementation also solves the issue https://issues.liferay.com/browse/LPP-38323.
So please let me know if there is anything I need to do.
Thank you.

**Comment by wincent** [comment]
Unless I am mistaken, I see no problem with `ci:forward`-ing this.

**Comment by wincent** [comment]
ci:forward

**Pull request closed without merging**